### PR TITLE
CC-263: Prevent retrying queries with a broken connection

### LINKED ISSUE #38: Looks like we are not handling some SQLExceptions correctly
I observed a case where disconnect from MySQL caused us to poll() in a tight loop, returning null every time. We should probably figure out unrecoverable SQL errors and either reconnect ourselves or kill the task so another worker can take over.

```
[2016-01-29 23:46:30,118] ERROR Failed to run query for table TimestampIncrementingTableQuerier{name='bar', query='null', topicPrefix='test-', timestampColumn='null', incrementingColumn='id'}: com.mysql.jdbc.exceptions.jdbc4.MySQLQueryInterruptedException: Query execution was interrupted (io.confluent.connect.jdbc.JdbcSourceTask)
[2016-01-29 23:46:30,122] ERROR Failed to run query for table TimestampIncrementingTableQuerier{name='bar', query='null', topicPrefix='test-', timestampColumn='null', incrementingColumn='id'}: com.mysql.jdbc.exceptions.jdbc4.CommunicationsException: Communications link failure

The last packet successfully received from the server was 9 milliseconds ago.  The last packet sent successfully to the server was 3 milliseconds ago. (io.confluent.connect.jdbc.JdbcSourceTask)
[2016-01-29 23:46:30,122] ERROR Failed to run query for table TimestampIncrementingTableQuerier{name='bar', query='null', topicPrefix='test-', timestampColumn='null', incrementingColumn='id'}: java.sql.SQLException: No operations allowed after statement closed. (io.confluent.connect.jdbc.JdbcSourceTask)
[2016-01-29 23:46:30,122] ERROR Failed to run query for table TimestampIncrementingTableQuerier{name='bar', query='null', topicPrefix='test-', timestampColumn='null', incrementingColumn='id'}: java.sql.SQLException: No operations allowed after statement closed. (io.confluent.connect.jdbc.JdbcSourceTask)
[2016-01-29 23:46:30,122] ERROR Failed to run query for table TimestampIncrementingTableQuerier{name='bar', query='null', topicPrefix='test-', timestampColumn='null', incrementingColumn='id'}: java.sql.SQLException: No operations allowed after statement closed. (io.confluent.connect.jdbc.JdbcSourceTask)
```

The last error repeats in tight loop forever.

### PR DESCRIPTION
**Pull request by shikhar**
Fixes #38

### CODE REVIEW

**Comment by shikhar** [comment]
Also need to address the `TableMonitorThread`.

**Comment by shikhar** [comment]
The failing tests are doing static mocking of `DriverManager.getConnection()`, which I'd like to fix up now that there is an easier path to do so. Should allow for enabling `reuseForks` for the test runner: https://github.com/confluentinc/kafka-connect-jdbc/blob/6b5a3abb2c62f06ad2598e33c221e03b92a666cb/pom.xml#L181-L182

**Comment by shikhar** [comment]
@confluentinc/connect for review

File: src/main/java/io/confluent/connect/jdbc/source/JdbcSourceTask.java [hunk 1]
```diff
@@ -171,56 +162,42 @@ public void stop() throws ConnectException {
...
-        long nextUpdate = querier.getLastUpdate() +
-                          config.getInt(JdbcSourceTaskConfig.POLL_INTERVAL_MS_CONFIG);
-        long untilNext = nextUpdate - now;
-        log.trace("Waiting {} ms to poll {} next", untilNext, querier.toString());
-        if (untilNext > 0) {
+        // If not in the middle of an update, wait for next update time
+        final long nextUpdate = querier.getLastUpdate() + config.getInt(JdbcSourceTaskConfig.POLL_INTERVAL_MS_CONFIG);
+        for (long untilNext = nextUpdate - time.milliseconds(); untilNext > 0; untilNext = nextUpdate - time.milliseconds()) {
+          // The purpose of the loop is to handle spurious wakeups
+          log.trace("Waiting {} ms to poll {} next", untilNext, querier.toString());
```
Comment (ewencp) [inline src/main/java/io/confluent/connect/jdbc/source/JdbcSourceTask.java]: Is `Thread.sleep` subject to spurious wakeups that aren't going to throw an exception? The javadocs don't seem to specify...

File: src/main/java/io/confluent/connect/jdbc/source/TableMonitorThread.java [hunk 2]
```diff
@@ -80,74 +85,70 @@ public void run() {
...
-      final List<String> tables;
-      try {
-        tables = JdbcUtils.getTables(db, schemaPattern, tableTypes);
-        log.debug("Got the following tables: " + Arrays.toString(tables.toArray()));
-      } catch (SQLException e) {
-        log.error("Error while trying to get updated table list, ignoring and waiting for next "
-                  + "table poll interval", e);
-        return false;
-      }
+  private synchronized boolean updateTables() {
```
Comment (ewencp) [inline src/main/java/io/confluent/connect/jdbc/source/TableMonitorThread.java]: Unrelated, but the comment is incomplete (which is almost definitely my fault -- I have a tendency to leave half completed sentences...)

File: src/main/java/io/confluent/connect/jdbc/source/JdbcSourceTask.java (hunk 1 above)
Comment (shikhar) [inline src/main/java/io/confluent/connect/jdbc/source/JdbcSourceTask.java]: Good question, the loop seems unnecessary, I will simplify.

**Pull request merged**